src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperMatchingConventions.cs

```diff
@@ -58,15 +58,29 @@ public static bool SatisfiesParentTag(ReadOnlySpan<char> parentTagNameWithoutPre
         return true;
     }
 
-    public static bool SatisfiesAttributes(IReadOnlyList<KeyValuePair<string, string>> tagAttributes, TagMatchingRuleDescriptor rule)
+    public static bool SatisfiesAttributes(ImmutableArray<KeyValuePair<string, string>> tagAttributes, TagMatchingRuleDescriptor rule)
     {
-        if (!rule.Attributes.All(
-            static (requiredAttribute, tagAttributes) => tagAttributes.Any(
-                static (attribute, requiredAttribute) => SatisfiesRequiredAttribute(attribute.Key, attribute.Value, requiredAttribute),
-                requiredAttribute),
-            tagAttributes))
+        var requiredAttributes = rule.Attributes;
+        var count = requiredAttributes.Count;
+
+        for (var i = 0; i < count; i++)
```

Comment:
Couldn't this be a `foreach`?

Reply:
Unfortunately, `TagMatchingRuleDescriptor.Attributes` is defined with an `IReadOnlyList`, so foreach will allocate an enumerator. 😢

Reply:
Sadness.

Reply:
Don't despair! I'll fix it once the compiler merge is in and I tackle https://github.com/dotnet/razor/issues/8554.